### PR TITLE
Add HeapRegionStateTable to support concurrent copy forward write barrier

### PR DESCRIPTION
**Pull request by youngar**

### CODE REVIEW

**Comment by youngar** [comment]
A couple of notes about these commits: We previously did not have any VLHGC code in OMR yet, so I have created a new `vlhgc` directory in the GC.  I also used the `OMR::GC` namespace for the class, and in the global namespace did `typedef OMR::GC::HeapRegionStateManager MM_HeapRegionStateManager`.  I decide to write a small test for this code, adding our first GC unit test.  To reduce the dependencies on this class, `HeapRegionStateTable::initialize()` takes an `MM_Forge` instead of an `MM_EnvironmentBase`. I'm hoping to ultimately remove `OMR_GC_VLHGC_CONCURRENT_COPY_FORWARD` build flag once all the new work is complete. I think this should be properly guarded to not affect OpenJ9 builds but I'd like to quickly test it out first.

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
Looks like enabling `OMR_GC_VLHGC `in OMR activates a NUMA test in the port library.  I'm going to fix the `#ifdef` and the compile failure.

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
Here is the openj9 side of these changes: https://github.com/youngar/openj9/pull/1

**Comment by charliegracie** [comment]
Can this be released without breaking any downstream projects? ie are changes like the OpenJ9 link provided required for downstream projects?

File: gc/base/vlhgc/HeapRegionStateTable.hpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+  protected:
+  private:
+};
+
+} // namespace GC
+} // namespace OMR
+
+typedef OMR::GC::HeapRegionStateTable MM_HeapRegionStateTable;
```
Comment (rwy0717) [inline gc/base/vlhgc/HeapRegionStateTable.hpp]: I actually wouldn't bother with this typedef.

File: gc/base/vlhgc/HeapRegionStateTable.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+void
+HeapRegionStateTable::tearDown(Forge *forge)
+{
+	if (_table != NULL) {
+		forge->free(_table);
+	}
+}
+
+} // namespace GC
+} // namespace OMR
```
Comment (rwy0717) [inline gc/base/vlhgc/HeapRegionStateTable.cpp]: missing newline

File: gc/base/vlhgc/HeapRegionStateTable.hpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+	MMINLINE uint8_t *getTable() { return _table; }
+
+	MMINLINE uintptr_t
+	getIndex(const void *heapAddress)
+	{
+		uintptr_t heapDelta = ((uintptr_t)heapAddress) - _heapBase;
+		uintptr_t index = heapDelta >> _regionShift; 
+		return index;
+	}
```
Comment (rwy0717) [inline gc/base/vlhgc/HeapRegionStateTable.hpp]: Is there any chance that an off-heap address comes through this API? You might want to keep the number of regions cached in the table, so you can assert the computed index is correct, at least initially.

File: gc/base/vlhgc/HeapRegionStateTable.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+bool
+HeapRegionStateTable::initialize(Forge *forge, uintptr_t heapBase, uintptr_t regionShift, uintptr_t regionCount)
+{
+	_heapBase = heapBase;
+	_regionShift= regionShift;
+
+	/* 1 byte per region */
+	_table = (uint8_t *) forge->allocate(regionCount, AllocationCategory::FIXED, OMR_GET_CALLSITE());
+	if (NULL == _table)
+	{
```
Comment (rwy0717) [inline gc/base/vlhgc/HeapRegionStateTable.cpp]: funky formatting

File: gc/base/GCExtensionsBase.hpp [hunk 5]
```diff
@@ -84,6 +85,14 @@ class MM_RealtimeGC;
 class MM_VerboseManagerBase;
 struct J9Pool;
 
+namespace OMR {
+namespace GC {
+#if defined(OMR_GC_VLHGC_CONCURRENT_COPY_FORWARD)
+class HeapRegionStateTable;
+#endif /* defined(OMR_GC_VLHGC_CONCURRENT_COPY_FORWARD) */
+} // namespace OMR
+} // namespace GC
```
Comment (rwy0717) [inline gc/base/GCExtensionsBase.hpp]: Why not just include the header?

File: gc/base/Forge.hpp [hunk 6]
```diff
@@ -67,8 +67,7 @@ friend class ::MM_GCExtensionsBase;
 	 * 
 	 */
 	void tearDown();
-	
-public:
+
```
Comment (rwy0717) [inline gc/base/Forge.hpp]: Did you do this so you could write the tests? It looks reasonable, I'm just wondering why. You should also get rid of the friend declaration above.

File: gc/base/GCExtensionsBase.hpp (hunk 5 above)
Comment (youngar) [reply gc/base/GCExtensionsBase.hpp]: As we talked about, including the VLHGC header in to GCExtensionsBase requires changing a lot of include paths throughout openj9.

File: gc/base/Forge.hpp (hunk 6 above)
Comment (youngar) [reply gc/base/Forge.hpp]: Yes for tests, friend class removed

**Comment by youngar** [comment]
@genie-omr build all

should be good now

**Comment by youngar** [comment]
@genie-omr build zos
System issue:
```
12:33:40  checking OMR_BUILD_TOOLCHAIN... Cannot contact ZISVJD04: hudson.remoting.RequestAbortedException: java.io.IOException: Unexpected termination of the channel
```

**Comment by charliegracie** [comment]
@rwy0717 are you ok with this change now?

**Comment by rwy0717** [comment]
yep, looks fine to me 👍

**Pull request merged**